journalctl: add --all-boots option

okdana opened the pull request:
**[RFC / DRAFT]**

When using a persistent journal, `journalctl` will by default show entries from previous boots in addition to the current one. I find this confusing and noisy in many cases, so i thought i would just add `-b0` to my shell alias for `journalctl`. But there is (AFAICT) no way to negate that option's effect afterwards, so whenever i actually *wanted* to see entries from all boots i would have to bypass my alias entirely (which is tedious and distracting).

So i simply want to teach `journalctl` an option to negate `-b`. I decided on `--all-boots`, because it seems safe and easy, but maybe something like `-b all` would make sense too?

Please let me know what you think — if it's viable, i will update the documentation, make sure the tests pass, &c., and come back to you.

keszybz commented:
I think that's reasonable. But I'd prefer `-b all', journalctl has way too many options already.

poettering commented:
yeah, "-ball" sounds better to me, too

okdana commented:
OK, updated and rebased.

* `parse_boot_descriptor()` now returns 1 for `all`, 0 for other legal inputs, and <0 for error

* In the course of adding `all`, i fixed an issue where using `-b` by itself would not always act like `-b0`. For example, if you did `journalctl -b -1 -b`, it would still have the effect of `-b -1`, in contradiction to the documentation

* I updated the documentation. Not sure if the formatting is OK. It looks like this when the man page is rendered:

  >**-b [[**_ID_][_±offset_]|**all**], **--boot[=[**_ID_][_±offset_]|**all**]

  The bold effect on the first two `[`s seems to be a bug in the conversion itself, because it affects other options in the documentation as well

* I added tests for `-b` == `-b0` and `-ball`, which pass if i perform them manually in the build directory, but i'm not sure how to run them as part of the test suite itself. `ninja test` passes, though

Does that seem alright?

poettering commented on src/journal/journalctl.c:
```diff
@@ -264,7 +264,9 @@ static int parse_boot_descriptor(const char *x, sd_id128_t *boot_id, int *offset
 
-        if (strlen(x) >= 32) {
+        if (streq(x, "all"))
+          return 1;
```
indentation. please follow https://systemd.io/CODING_STYLE, i.e. 8ch indenting

poettering replied:
we usually follow the rule that functions shall never clobber output parameters on error, and always initialize them on success. Given that this is a success situation (i.e. it returns >= 0, rather than < 0) it hence should reset `*boot_id`, and `*offset` in that case.

If you do that, the calling side becomes a bit simpler too, as you don't have to initialize those two vars explicitly first.

poettering replied:
Also, I think it would be nicer to make the return value be == 0 in case of "all", and > 0 in case a normal boot descriptor is specified. That way the return value would map directly to arg_boot later on (except for the < 0 error case).

poettering commented on src/journal/journalctl.c:
```diff
@@ -611,10 +617,12 @@ static int parse_argv(int argc, char *argv[]) {
                                 if (optind < argc &&
-                                    parse_boot_descriptor(argv[optind], &arg_boot_id, &arg_boot_offset) >= 0)
+                                    (r = parse_boot_descriptor(argv[optind], &arg_boot_id, &arg_boot_offset)) >= 0)
                                         optind++;
```
let's not pack so much stuff into the if expression. please do this instead:

```
if (optarg) {
        r = parse_boot_descriptor(optarg, &arg_boot_id, &arg_boot_offset);
        if (r < 0) 
                return log_error_errno(r, "Failed to parse boot descriptor '%s'", optarg);

        arg_boot = r;
} else if (optind < argc) {
        r = parse_boot_descriptor(…);
        if (r >= 0) {
                arg_boot = r;
                optind++;
        }
}
```

poettering commented:
Also, CI failed badly on this one

okdana commented:
Thanks!

I altered your suggestions regarding the `-b` logic slightly:

1. I added an `else arg_boot = true` after your suggested `else if`, since otherwise `-b` without argument has no effect.

2. Regarding your comment that i 'don't have to initialize those two vars explicitly first', AFAICT it's necessary to re-initialise these somewhere in the `-b` logic (whether at the top or inside my added `else`) to fix the issue where `-b`-without-arg doesn't behave like `-b0` if you've used `-b`-with-arg previously. (I realise that's technically a separate issue from the one at hand, but...)

Does that make sense?

(Re: the CI failures, i noticed those too, but none of them seemed related to my changes)

keszybz commented on src/journal/journalctl.c:
```diff
@@ -264,7 +264,11 @@ static int parse_boot_descriptor(const char *x, sd_id128_t *boot_id, int *offset
 
-        if (strlen(x) >= 32) {
+        if (streq(x, "all")) {
+                *boot_id = id;
```
Eh, just say `*boot_id = SD_ID128_NULL; *offset = 0;`, that much more obvious to the reader, and much less likely to be messed up by some subsequent refactoring.

keszybz commented on src/journal/journalctl.c:
```diff
@@ -591,30 +595,35 @@ static int parse_argv(int argc, char *argv[]) {
 
                 case 'b':
-                        arg_boot = true;
+                        r = 0;
```
This initialization is not needed for anything afaics.

okdana commented:
Oops, fixed. Thanks!

poettering commented on src/journal/journalctl.c:
```diff
@@ -591,30 +595,34 @@ static int parse_argv(int argc, char *argv[]) {
                 case 'b':
-                        arg_boot = true;
+                        arg_boot_id = SD_ID128_NULL;
+                        arg_boot_offset = 0;
```
please move these lines into the final, `arg_boot = true` branch of the if statement. After all, if there's success in the first two branches, then these fields are reset anyway. And if they fail, then there's no point in overriding these fails. Thus only in the third branch case it makes sense to reset this.

okdana replied:
I was going to make this change too, but when i went back to it i realised that the `else` thing is flawed — it leaves unaccounted-for the condition where `-b` is used, and there is a following argument, but it isn't a boot descriptor, as in `journalctl -ball -b -q`. In that case, `-b` will do nothing at all again (rather than act like `-b0`). So i moved all of the re-initialisation stuff back up to the top (and i added a test to catch that case i'd initially missed). Please let me know if you think there's a better way after all

keszybz commented on test/TEST-04-JOURNAL/test-journal.sh:
```diff
@@ -63,6 +63,14 @@ grep -q '^PRIORITY=6$' /output
+cmp /expected /output
+
+# `-b all` negates earlier use of -b (-b and -m are otherwise exclusive)
+journalctl -b -1 -b all -m
```
So, this doesn't actually test anything.
What about the following
```
journalctl -m | head -1 >/expected
journalctl -b 56 -b all | head -1 >/output
cmp /expected /output
```

okdana replied:
Hm, maybe i'm misunderstanding. `journalctl -b -1 -b all -m` is testing the fact that `-ball` negates the previous use of `-b` — if it did not, the command would fail with the following error:

>Using --boot or --list-boots with --merge is not supported.

Is that not a valid test?

If not, what is the `-m` doing for us in your suggestion? I assume the test environment doesn't actually have any other journals to merge — and if it did, wouldn't it break the comparison?

okdana replied:
(As mentioned in my reply to Lennart, i added another test that makes use of `-m` in this fashion, to confirm what effect `-b` is having. Let me know if you see any issues with it)

keszybz commented:
LGTM.
